src/irmin-pack/inode.ml

```diff
@@ -1687,6 +1687,129 @@ struct
       let of_concrete t = proof_of_concrete (lazy (failwith "blinded root")) t
       let to_concrete = concrete_of_proof ~depth:0
     end
+
+    module Snapshot = struct
+      type 'a val_impl = 'a t
+      type kinded_hash = Contents of hash | Node of hash [@@deriving irmin]
+      type entry = { name : step; hash : kinded_hash } [@@deriving irmin]
+
+      type inode_tree = {
+        depth : int;
+        length : int;
+        pointers : (int * hash) list;
+      }
+      [@@deriving irmin]
+
+      type v = Inode_tree of inode_tree | Inode_value of entry list
+      [@@deriving irmin]
+
+      type t = { v : v; root : bool } [@@deriving irmin]
+
+      let get_key ~index h =
+        match index h with
+        | None ->
+            Fmt.failwith
+              "get_key: You are trying to save to the backend an inode \
+               deserialized using [Irmin.Type] that used to contain pointer(s) \
+               to inodes which are unknown to the backend. Hash: %a"
+              pp_hash h
+        | Some key -> key
+
+      let get_hash k = Key.to_hash k
+
+      let to_entry (name, v) =
+        match v with
+        | `Contents (contents_key, _) ->
+            let h = get_hash contents_key in
+            { name; hash = Contents h }
+        | `Node node_key ->
+            let h = get_hash node_key in
+            { name; hash = Node h }
+
+      let to_pred_entry (_, v) =
+        match v with
+        | `Contents (contents_key, _) ->
+            let h = get_hash contents_key in
+            `Contents h
+        | `Node node_key ->
+            let h = get_hash node_key in
+            `Node h
+
+      let of_entry ~index e : step * Node.value =
+        ( e.name,
+          match e.hash with
+          | Contents hash ->
+              let key = get_key ~index hash in
+              `Contents (key, Metadata.default)
```

Comment:
I think that the snapshot should contain the metadata, even if we end up using `unit` in lib_context